### PR TITLE
Use Rcpp for 10x faster data.frame conversion

### LINKED ISSUE #33: R client much slower than Python client
Hi, I found a python client for Presto at

https://github.com/easydatawarehousing/prestoclient/blob/master/python/prestoclient.py

On a table with 1.7 million rows this client is much faster than RPresto. Digging around the code base I found dbFetch is implemented slightly inefficiently in https://github.com/prestodb/RPresto/blob/master/R/dbFetch.R

The R function creates an empty data frame, and then loops through the result set calling 'rbind'. This is slow because R will allocate memory for the new rv data frame, and then on the next iteration it will resize etc. It's easier to store the results in a list, and then use do.call('rbind', rv) at the end to collapse the results into 1 data frame.

dbBigQuery = function(con, sql) {
  send = dbSendQuery(con, sql)
  results = list()
  counter = 1
  while (!dbHasCompleted(send)) {
    tryCatch({
      chunk <- dbFetch(send)
      print (sprintf("downloaded %s rows", nrow(chunk)))
      results[[counter]] = chunk
      counter = counter + 1  
    }, error = function(e) {
      message("encountered error: will skip this check and continue downloading results")
      message(e)
    })
  }
  return (do.call('rbind', results))
}

After this simple speedup I got results from my 1.7M row table in 12-13 mins instead of 15. However, if I were to use the Python client and dump the result to a text file for R read in, I can get the results in < 5 mins. Why is the Python client so much faster? They are both hitting the  <host>/v1/statement endpoint

The python client hasn't been maintained for over a year, but still seems to work. Wondering if it's a smarter strategy for the RPresto library to just wrap the python client? Would this be a solution you are willing to adopt ?

### PR DESCRIPTION
**Pull request by saurfang**
Move row iterations from R to Rcpp for 10x speedup. `check_names` verify column names are consistent and each row contains exactly required columns. `transpose` convert a row major list of list to a column major list of list. It also handles null to NA conversion for inner lists and special numerical values. Because assignment is made in-place whenever possible, this is much faster than the replaced R implementation.

Fix #33 hopefully. I not well-versed in Rcpp so feedbacks and suggestions are really appreciated. Initial profiling result seems promising, for exact same query in #57:

image

`.extract.data` nows takes about the same time as raw JSON parsing and is 10x faster compared to original runtime and profiler credits `.tranpose` 20ms out of 110ms (note 10ms is minimal resolution). 
I still need to take a look at bigint and add some test cases for named/unnamed list but feel free to try this out and let me know if this fixes the performance issue or if it outputs garbage.

### CODE REVIEW

**Comment by onurfiliz** [comment]
This is great work @saurfang, thank you. @cosinequanon has been looking into a similar approach, so he may be able to speak better to the approach.

Just to verify, do the unit tests pass via `devtools::test`?

**Comment by cosinequanon** [comment]
Oh this is so so awesome! Thank you so much for doing this.

I have a commit that rewrites the json.tabular.to.data.frame.R file to be written in Rcpp but it only works for basic (int/numeric/character/boolean) types. We ran into some issues with making all the datetime timezone stuff work correctly. I can share the code in a gist if you would find that useful but to be honest it does not past the tests. I just haven't found the time to iron out all the final kinks.

If your code passes the tests then we would be happy to do whatever needs to be done to get it merged. I'm not super good with C++ tbh but I know some people that know the language well and could help review code.

**Comment by saurfang** [comment]
Cool! Those tests that don't need a Presto connection did pass for me. I haven't got the time to set up a local Presto for rest of the tests. I am pretty certain I didn't cover all the basis for corner cases on numerical and integer types but shouldn't be a big issue once I create unit tests for them. If you can share your progress in a Gist, that'd be great and I can see if there is anything I can borrow.
Let me spend a bit more time later this week and will appreciate some C++ code review then.

**Comment by cosinequanon** [comment]
Here's a gist with some WIP code, should give you a sense for what I was trying to do https://gist.github.com/cosinequanon/23812c4d949434ab2222f54ffd17b8df

Feel free to take anything and add to this PR, I'm not tied to my code, I just want to see it live ;)

That said, we do need to test this on a live instance, it would be useful for you to get one setup so you can make sure everything passes. We will patch this code and make sure the tests pass against our instance before accepting, I know we (unfortunately) have a fair amount of unit tests that rely on Presto being available.

**Comment by cosinequanon** [comment]
Let us know when you are ready for a review and I'll take a closer look at the code.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by saurfang** [comment]
Okay good to go. All tests passed. I feel like `transpose` can use some DRY but not sure how. Everything left on the R side is pretty vectorized and doesn't seem to cause any bottleneck. Let me know what you think.

p.s. I couldn't find specific instructions to setup presto for test. So I forked a docker setup for the live connection testing for future reference: https://github.com/saurfang/docker-rpresto

**Comment by cosinequanon** [comment]
I pulled your changes locally but has an issue compiling, this is the error I saw https://gist.github.com/cosinequanon/a137b037f4c4460363d3bb3b0967659d I'm using gcc version 4.8.5 any idea what might be the issue?

**Comment by facebook-github-bot-8** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by saurfang** [comment]
hmm. Apparently `std::to_string` is C++11 only. I replaced it with Rcpp variants. Can you give it another try?

File: src/check_names.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// jsonify a character vector
+std::string jsonify(CharacterVector x) {
+  std::string out = "[";
+  for (int i = 0; i < x.size(); i++) {
+    out += "\"" + x[i] + "\",";
+  }
+  // trim trailing comma
+  if (x.size() > 0) {
+    out = out.substr(0, out.size() - 1);
```
Comment (czarandy) [inline src/check_names.cpp]: This copies the entire string. You could just do out.resize(out.size() - 1) instead which should be much more efficient.
Alternatively, you can append the first element and then loop over the remaining elements appending the separator first.

File: src/check_names.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// jsonify a character vector
+std::string jsonify(CharacterVector x) {
+  std::string out = "[";
+  for (int i = 0; i < x.size(); i++) {
+    out += "\"" + x[i] + "\",";
```
Comment (czarandy) [inline src/check_names.cpp]: Using += is more efficient because it avoids extra copies, so:
out += "\"";
out += x[i];
out += "\",";

would be the preferred way

**Comment by cosinequanon** [comment]
Thanks @czarandy for the review (he is the C++ expert I mentioned) I'm about to give it another shot for compiling

**Comment by cosinequanon** [comment]
Okay so all the tests pass for me, I checked the perf on our data and found that for a "wide" query (100+ columns, 10k rows) we have a 2x improvement and for a long query (2 columns, 1M rows) we have a 4x improvement. We should most definitely ship this, I'd like @onurfiliz to review this as well before merging.

**Comment by facebook-github-bot-3** [comment]
@saurfang updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@saurfang updated the pull request.

**Comment by saurfang** [comment]
Good to know. Thanks a lot for the reviews @czarandy 

@cosinequanon I have addressed suggestions and retested the package. I also ran some profiling and looks like `.extract.data` takes about 10-20% runtime (which is about half the time spent in `jsonlite::fromJSON`)

**Comment by onurfiliz** [comment]
@saurfang thanks a lot for all your work, this is great. We may need to go over the test suite to make sure the new code path is tested. In general things seem good, but it may still take me 7-10 days to review this, since it is a rather extensive change. Just wanted to let you know, so that you don't think it slipped through the cracks ;)

**Comment by saurfang** [comment]
@onurfiliz I appreciate your commitment to correctness and good test coverage. Let me know if you have any questions or comments.

**Comment by onurfiliz** [comment]
I have gone over the diff. In general it looks great, thank you. I will add some nits inline, mostly around using Rcpp API instead of C, after which it should be good to go.

File: R/json.tabular.to.data.frame.R [hunk 3]
```diff
@@ -6,20 +6,10 @@
 # of patent rights can be found in the PATENTS file in the same directory.
 
 #' @include dbDataType.R
+#' @useDynLib RPresto
+#' @importFrom Rcpp sourceCpp
```
Comment (onurfiliz) [inline R/json.tabular.to.data.frame.R]: Is this needed?

File: src/check_names.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// jsonify a character vector
+std::string jsonify(CharacterVector x) {
```
Comment (onurfiliz) [inline src/check_names.cpp]: Since this is hopefully not going to be called very often (if at all), it's not performance sensitive. It may be better to use base::paste instead, like:

    Function paste = Environment("package:base")["paste"];
    return paste(x, Named("collapse", ", "));

File: src/check_names.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  for (int i = 0; i < x.size(); i++) {
+    List row = x[i];
+
+    if (row.size() != column_count) {
+      stop("Item " + toString(i) + ", " +
+           "expected: " + toString(column_count) + " columns, " +
+           "received: " + toString(row.size()));
+    }
+
+    if (!Rf_isNull(row.names())) {
```
Comment (onurfiliz) [inline src/check_names.cpp]: row.names().isNULL()

may be more canonical.

File: src/transpose.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      column[r] = as<RTYPE>(cell);
+    }
+  }
+}
+
+// set null element in list to NA (NA_LOGICAL) recursively
+void null_to_na(List x) {
+  int size = x.size();
+  for (int i = 0; i < size; i++) {
+    SEXP elem = x[i];
```
Comment (onurfiliz) [inline src/transpose.cpp]: RObject element = x[i];

(I have an obsession against abbreviations:))

File: src/transpose.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    if (elem == R_NilValue) {
```
Comment (onurfiliz) [inline src/transpose.cpp]: element.isNULL()

File: src/transpose.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      x[i] = LogicalVector::create(NA_LOGICAL);
+    } else if(TYPEOF(elem) == VECSXP) {
```
Comment (onurfiliz) [inline src/transpose.cpp]: element.sexp_type() == VECSXP

File: src/transpose.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+// tranpose a nested list
+// e.g. list(a=list(x=1,y=2,z=3), b=list(x=1,y=2,z=3)) =>
+//        list(x=list(a=1,b=1), y=list(a=2,b=2), z=list(a=3,b=3))
+// x is the list to tranpose
+// out is a pre-allocated correctly-typed output placeholder
+// type coercion is applied according to `out` type as necessary
+// [[Rcpp::export(name = ".transpose")]]
+List transpose(List x, List out) {
+  int rows = x.size();
```
Comment (onurfiliz) [inline src/transpose.cpp]: Could you use

    int row_count = x.size();
    int column_count = out.size();

instead?

File: src/transpose.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+// tranpose a nested list
+// e.g. list(a=list(x=1,y=2,z=3), b=list(x=1,y=2,z=3)) =>
+//        list(x=list(a=1,b=1), y=list(a=2,b=2), z=list(a=3,b=3))
+// x is the list to tranpose
+// out is a pre-allocated correctly-typed output placeholder
+// type coercion is applied according to `out` type as necessary
+// [[Rcpp::export(name = ".transpose")]]
+List transpose(List x, List out) {
```
Comment (onurfiliz) [inline src/transpose.cpp]: List transpose(List input, List output) {

File: src/transpose.cpp [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  int rows = x.size();
+  if(rows == 0) return out;
```
Comment (onurfiliz) [inline src/transpose.cpp]: Let's add braces:

    if (row_count == 0) {
      return output;
    }

File: src/transpose.cpp [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+  int cols = out.size();
+  if(cols == 0) return out;
+
+  for (int c = 0; c < cols; c++) {
+    SEXP column = VECTOR_ELT(out, c);
```
Comment (onurfiliz) [inline src/transpose.cpp]: RObject column = output[c];

File: src/transpose.cpp [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    switch(TYPEOF(column)) {
```
Comment (onurfiliz) [inline src/transpose.cpp]: Maybe convert all:

- `TYPEOF` -> `.sexp_type()`
- `== R_NilValue` -> `.isNULL()`
- `SEXP x = VECTOR_ELT` -> `RObject x = a[..]`
- `SET_VECTOR_ELT` -> `a[..] = x`

File: src/transpose.cpp [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+          SET_VECTOR_ELT(column, r, cell);
+        }
+      }
+      break;
+    }
+    case INTSXP: {
+      // keep track whether we encounter a broken integer
+      bool broken_int = false;
+      // resulting vector if all integers are good
+      IntegerVector intColumn = as<IntegerVector>(column);
```
Comment (onurfiliz) [inline src/transpose.cpp]: Use `_` instead of camel case in the whole file.

File: src/transpose.cpp [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            null_to_na(cell);
+          }
+          SET_VECTOR_ELT(column, r, cell);
+        }
+      }
+      break;
+    }
+    case INTSXP: {
+      // keep track whether we encounter a broken integer
+      bool broken_int = false;
```
Comment (onurfiliz) [inline src/transpose.cpp]: I'd use `broken_integer` or `large_integer`, and add a comment as to why this would happen.

File: src/transpose.cpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+  for (int c = 0; c < cols; c++) {
+    SEXP column = VECTOR_ELT(out, c);
+    switch(TYPEOF(column)) {
+    case VECSXP: { // list
+      for (int r = 0; r < rows; r++) {
+        SEXP row = VECTOR_ELT(x, r);
+        SEXP cell = VECTOR_ELT(row, c);
+        if (cell != R_NilValue) {
+          if(TYPEOF(cell) == VECSXP) {
+            null_to_na(cell);
```
Comment (onurfiliz) [inline src/transpose.cpp]: Hmm, note that this line makes the new situation semantically different from the previous pure R code. We now could possibly be modifying the input, which I'm not crazy about. On the other hand, I can see the benefits of in-place modification. In either case, this should be documented in a comment somewhere, probably right above the `transpose` call in `json.tabular.to.data.frame.R`.

File: src/transpose.cpp [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+      IntegerVector intColumn = as<IntegerVector>(column);
+      // placeholder vector if we need to coerce into numerics
+      NumericVector numColumn;
+
+      for (int r = 0; r < rows; r++) {
+        List row = x[r];
+        SEXP cell = row[c];
+        if (cell != R_NilValue) {
+          // keep populating integer vector
+          if (!broken_int && TYPEOF(cell) == INTSXP) {
```
Comment (onurfiliz) [inline src/transpose.cpp]: I found this block a bit hard to read. I have an alternate representation which I feel is more expressive, feel free to keep it as is if you disagree:

    if (broken_int) {
      numColumn[r] = as<double>(cell);
    } else { // not broken yet
      if (cell.sexp_type() == INTSXP) {
        intColumn[r] = as<int>(cell);
      } else { // we found the first item that is broken
        numColumn = as<NumericVector>(intColumn);
        numColumn[r] = as<double>(cell);
        broken_int = true;
      }
    }

File: src/transpose.cpp [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+              numColumn = as<NumericVector>(intColumn);
+              broken_int = true;
+            }
+            numColumn[r] = as<double>(cell);
+          }
+        }
+      }
+
+      // set to numeric vector for bigint coercion
+      out[c] = broken_int ? numColumn : intColumn;
```
Comment (onurfiliz) [inline src/transpose.cpp]: Don't we just need?:

    if (broken_int) {
      out[c] = numColumn;
    }

File: src/transpose.cpp [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+      break;
+    }
+    case REALSXP: {
+      // handle special cases
+      NumericVector tmp = as<NumericVector>(column);
```
Comment (onurfiliz) [inline src/transpose.cpp]: Could you rename `tmp` to something more representative? `tmp` in my mind implies throw-away, whereas this is still directly accessing `output`.

File: src/transpose.cpp [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+      for (int r = 0; r < rows; r++) {
+        List row = x[r];
+        SEXP cell = row[c];
+        if (cell != R_NilValue) {
+          if (TYPEOF(cell) == STRSXP) {
+            String cellStr = as<String>(cell);
```
Comment (onurfiliz) [inline src/transpose.cpp]: `cell_string`

File: src/transpose.cpp [hunk 21]
```diff
@@ -0,0 +1,140 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// copy from x with number of rows, to c column in columns
+template <typename T, class RTYPE>
```
Comment (onurfiliz) [inline src/transpose.cpp]: This is not `RTYPE` but rather `CPPTYPE`?

I'm also not crazy about the name `copy`, but I don't have a better suggestion off the top of my mind.

File: src/transpose.cpp [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    case LGLSXP: {
+      copy<LogicalVector, bool>(as<LogicalVector>(column), rows, x, c);
+      break;
+    }
+    case STRSXP: {
+      copy<CharacterVector, String>(as<CharacterVector>(column), rows, x, c);
+      break;
+    }
+    default: {
```
Comment (onurfiliz) [inline src/transpose.cpp]: Do we not get `RAWSXP` from jsonlite?

File: R/json.tabular.to.data.frame.R (hunk 3 above)
Comment (saurfang) [inline R/json.tabular.to.data.frame.R]: Seems so so that Roxygen can generate NAMESPACES properly.
https://github.com/hadley/devtools/blob/01ab167ca8663562c30a2524d3f2907b2c2ba10d/R/infrastructure.R#L125

File: src/check_names.cpp (hunk 5 above)
Comment (saurfang) [inline src/check_names.cpp]: I don't believe this works because `row.names()` returns a `NamesProxy` object. Also see http://stackoverflow.com/questions/24744802/rcpp-how-to-check-if-any-attribute-is-null

File: src/transpose.cpp (hunk 16 above)
Comment (saurfang) [inline src/transpose.cpp]: That's true. I added comments in both function to emphasize null will be replaced in-place.

File: src/transpose.cpp (hunk 21 above)
Comment (saurfang) [inline src/transpose.cpp]: Felt the same. `copy_column` or `pluck_column`? I'm going with `copy_column` for now.

File: src/transpose.cpp (hunk 22 above)
Comment (saurfang) [inline src/transpose.cpp]: I believe binary types are passed base64 encoded from the REST API: https://github.com/prestodb/RPresto/blob/f1f722addab4528b88847a387be128ebf392dc30/R/json.tabular.to.data.frame.R#L127

**Comment by saurfang** [comment]
@onurfiliz thanks for the review! I have addressed most of your comments. Let me know if there anything else. I reran the test suite and they seem to pass fine.

@facebook-github-bot welcome! I like your avatar.

**Comment by onurfiliz** [comment]
@saurfang excellent, thank you for your awesome contribution.

**Comment by onurfiliz** [comment]
To close the loop on this, I've run some comparisons across different branches, and the results look solid:

timing_comparison

Here 'base' stands for the state of things before the `null_to_na` PR, and `cli` stands for our internal wrapper around the presto CLI's csv output that just shells out via `system2`. I have replicated each run 10 times. We get around 3x-7x improvement from 'base', and are faster than the CLI hack (which has no type safety) up until 150M cells or so.

**Comment by saurfang** [comment]
woot. Thanks a lot for following up @onurfiliz!
